wsmaster/che-core-api-devfile/src/main/java/org/eclipse/che/api/devfile/server/jpa/JpaUserDevfileDao.java

```diff
@@ -0,0 +1,420 @@
+      return managerProvider
+          .get()
+          .createNamedQuery("UserDevfile.getTotalCount", Long.class)
+          .getSingleResult();
+    } catch (RuntimeException x) {
+      throw new ServerException(x.getLocalizedMessage(), x);
+    }
+  }
+
+  @Transactional
+  protected void doCreate(UserDevfileImpl devfile) {
+    final EntityManager manager = managerProvider.get();
+    manager.persist(devfile);
+    manager.flush();
+  }
+
+  @Transactional
+  protected Optional<UserDevfileImpl> doUpdate(UserDevfileImpl update) {
+    final EntityManager manager = managerProvider.get();
+    if (manager.find(UserDevfileImpl.class, update.getId()) == null) {
+      return Optional.empty();
+    }
+    UserDevfileImpl merged = manager.merge(update);
+    manager.flush();
+    return Optional.of(merged);
+  }
+
+  @Transactional(rollbackOn = {RuntimeException.class, ServerException.class})
+  protected void doRemove(String id) throws ServerException {
+    final EntityManager manager = managerProvider.get();
+    final UserDevfileImpl devfile = manager.find(UserDevfileImpl.class, id);
+    if (devfile != null) {
+      eventService
+          .publish(new BeforeDevfileRemovedEvent(new UserDevfileImpl(devfile)))
+          .propagateException();
+      manager.remove(devfile);
+      manager.flush();
+    }
+  }
+
+  public static class UserDevfileSearchQueryBuilder {
+    protected EntityManager entityManager;
+    protected int maxItems;
+    protected int skipCount;
+    protected String filter;
+    protected Map<String, String> params;
+    protected String order;
+
+    public UserDevfileSearchQueryBuilder(EntityManager entityManager) {
+      this.entityManager = entityManager;
+      this.params = new HashMap<>();
+    }
+
+    public static UserDevfileSearchQueryBuilder newBuilder(EntityManager entityManager) {
+      return new JpaUserDevfileDao.UserDevfileSearchQueryBuilder(entityManager);
+    }
+
+    public UserDevfileSearchQueryBuilder withMaxItems(int maxItems) {
+      this.maxItems = maxItems;
+      return this;
+    }
+
+    public UserDevfileSearchQueryBuilder withSkipCount(int skipCount) {
+      this.skipCount = skipCount;
+      return this;
+    }
+
+    public UserDevfileSearchQueryBuilder withFilter(List<Pair<String, String>> filter) {
+      if (filter == null || filter.isEmpty()) {
+        this.filter = "";
+        return this;
+      }
+      final StringJoiner matcher = new StringJoiner(" AND ", " WHERE ", " ");
+      int i = 0;
+      for (Pair<String, String> attribute : filter) {
+
+        final String parameterName = "parameterName" + i++;
+        if (attribute.second.startsWith("like:")) {
+          params.put(parameterName, attribute.second.substring(5));
+          matcher.add("userdevfile." + attribute.first + " LIKE :" + parameterName);
+        } else {
+          params.put(parameterName, attribute.second);
+          matcher.add("userdevfile." + attribute.first + " = :" + parameterName);
```

Comment:
We do the validation of the attributes upfront but maybe it should be noted somewhere that the `attribute.first` is put into the query further unsanitized. If we fail to remember that in the future and allow for "arbitrary" keys, we run in the risk of SQL injection.